README.md
# Swipperview 🎉

<img src="https://github.com/rkumar1904/swipperview/raw/master/assets/vid.gif" height="300" />

## Getting Started

```dart
  body: Swipper(
        data: productList,
        icon: Icons.headphones,
        onChange: (page) {
          print('swipped $page');
        },
        actionName: 'Buy Now',
        actionHandler: (index) {
          print('action Click $index');
        },
      ),
```

```dart
List<Map<String, dynamic>> productList = [
    {
      'title': '#1 Product of the Day',
      'description':
          '#1 This project is a starting point for a Dart package, a library module containing code that can be shared easily across multiple Flutter or Dart projects.',
      'url':
          'https://images.unsplash.com/photo-1611930022073-b7a4ba5fcccd?ixid=MnwxMjA3fDB8MHxzZWFyY2h8M3x8cHJvZHVjdHN8ZW58MHwxfDB8fA%3D%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=800&q=60',
    },
    {
      'title': '#2 Product of the Day',
      'description':
          '#2 This project is a starting point for a Dart package, a library module containing code that can be shared easily across multiple Flutter or Dart projects.',
      'url':
          'https://images.unsplash.com/photo-1610824352934-c10d87b700cc?ixid=MnwxMjA3fDB8MHxzZWFyY2h8MjV8fHByb2R1Y3RzfGVufDB8MXwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=800&q=60',
    },
    {
      'title': '#3 Product of the Day',
      'description':
          '#3 This project is a starting point for a Dart package, a library module containing code that can be shared easily across multiple Flutter or Dart projects.',
      'url':
          'https://images.unsplash.com/photo-1525966222134-fcfa99b8ae77?ixid=MnwxMjA3fDB8MHxzZWFyY2h8MzJ8fHByb2R1Y3RzfGVufDB8MXwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=800&q=60',
    },
    {
      'title': '#4 Product of the Day',
      'description':
          '#4 This project is a starting point for a Dart package, a library module containing code that can be shared easily across multiple Flutter or Dart projects.',
      'url':
          'https://images.unsplash.com/photo-1556228720-195a672e8a03?ixid=MnwxMjA3fDB8MHxzZWFyY2h8NTR8fHByb2R1Y3RzfGVufDB8MXwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=800&q=60',
    },
  ];

```



<img src="https://github.com/rkumar1904/swipperview/raw/master/assets/1.png?raw=true" height="400" />
<img src="https://github.com/rkumar1904/swipperview/raw/master/assets/2.png?raw=true" height="400" />



This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our
[online documentation](https://flutter.dev/docs), which offers tutorials,
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## 0.0.2+11

Added Example code and modfied sample files.

## 0.0.1+3

Added Sample code with less dependency

## 0.0.1+2

added custom view.


test/swipperview_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:swipperview/swipperview.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


example/swipper_example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:swipper_example/main.dart';

void main() {
  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(const MyApp());

    // Verify that our counter starts at 0.
    expect(find.text('0'), findsOneWidget);
    expect(find.text('1'), findsNothing);

    // Tap the '+' icon and trigger a frame.
    await tester.tap(find.byIcon(Icons.add));
    await tester.pump();

    // Verify that our counter has incremented.
    expect(find.text('0'), findsNothing);
    expect(find.text('1'), findsOneWidget);
  });
}


lib/src/swipper.dart
import 'package:flutter/material.dart';
import 'dart:math';
import 'package:palette_generator/palette_generator.dart';

class Swipper extends StatefulWidget {
  final ValueChanged<int>? onChange;
  final String actionName;
  final ValueChanged<int>? actionHandler;
  final IconData icon;

  final List<Map<String, dynamic>> data;
  const Swipper({
    Key? key,
    required this.onChange,
    required this.actionName,
    required this.actionHandler,
    required this.icon,
    required this.data,
  }) : super(key: key);

  @override
  _SwipperState createState() => _SwipperState();
}

class _SwipperState extends State<Swipper> {
  PageController controller = PageController(initialPage: 0);
  var imgURL = 'https://source.unsplash.com/collection/345710/640x960?' +
      Random().nextInt(1000).toString();
  bool isDarkColor = false;
  Color dynamicColor = Color(0xfff2f2f2);
  Future<Color> getImagePalette(ImageProvider imageProvider) async {
    final PaletteGenerator paletteGenerator =
        await PaletteGenerator.fromImageProvider(imageProvider);

    if (paletteGenerator.darkVibrantColor == null) {
      setState(() {
        isDarkColor = false;
      });
      return paletteGenerator.dominantColor!.color;
    } else {
      setState(() {
        isDarkColor = true;
      });
      return paletteGenerator.darkVibrantColor!.color;
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black,
      extendBodyBehindAppBar: true,
      appBar: AppBar(
        backgroundColor: Colors.transparent,
        elevation: 0,
      ),
      body: PageView.builder(
        controller: controller,
        scrollDirection: Axis.vertical,
        itemCount: widget.data.length,
        onPageChanged: (currentPage) async {
          print('current page $currentPage');
          // var url = 'https://source.unsplash.com/collection/345710/' +
          //     Random().nextInt(1000).toString();
          Color cc = await getImagePalette(
              NetworkImage(widget.data[currentPage]['url']));
          setState(() {
            imgURL = widget.data[currentPage]['url'].toString();
            dynamicColor = cc;
          });
          // widget.onChange!(currentPage);
        },
        itemBuilder: (context, index) {
          return Container(
            decoration: BoxDecoration(
              color: Colors.white,
              image: DecorationImage(
                fit: BoxFit.cover,
                image: NetworkImage(widget.data[index]['url'].toString()),
                // colorFilter: ColorFilter.linearToSrgbGamma(),
              ),
            ),
            child: Container(
                // height: 300,
                padding: const EdgeInsets.all(20),
                decoration: BoxDecoration(
                  gradient: LinearGradient(
                    // begin: Alignment.bottomCenter,
                    begin: Alignment(0.0, 0.0),
                    end: Alignment(0.0, 0.3),
                    colors: <Color>[
                      dynamicColor.withOpacity(0.1),
                      dynamicColor,
                      // priamary
                      // Colors.blue.withOpacity(0.1),
                      // Colors.blue,
                    ],
                    tileMode: TileMode.clamp,
                  ),
                ),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Expanded(
                      flex: 3,
                      child: Container(),
                    ),
                    Expanded(
                      flex: 2,
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        mainAxisAlignment: MainAxisAlignment.spaceAround,
                        children: [
                          Icon(
                            widget.icon,
                            size: 50,
                            color:
                                isDarkColor ? Colors.white54 : Colors.black54,
                          ),
                          // widget.icon,

                          Text(
                            widget.data[index]['title'].toString(),
                            style: TextStyle(
                              fontSize: 25,
                              color:
                                  isDarkColor ? Colors.white54 : Colors.black54,
                              fontWeight: FontWeight.w600,
                            ),
                          ),
                          Text(
                            widget.data[index]['description'].toString(),
                            style: TextStyle(
                              fontSize: 18,
                              color:
                                  isDarkColor ? Colors.white54 : Colors.black54,
                              fontWeight: FontWeight.w600,
                            ),
                          ),
                          Center(
                            child: ElevatedButton.icon(
                              onPressed: () {
                                widget.actionHandler!(index);
                              },
                              icon: Icon(
                                Icons.shopping_bag_outlined,
                                color: isDarkColor
                                    ? Colors.white54
                                    : Colors.black54,
                              ),
                              label: Text(widget.actionName,
                                  style: TextStyle(
                                    color: isDarkColor
                                        ? Colors.white54
                                        : Colors.black54,
                                  )),
                              style: ElevatedButton.styleFrom(
                                backgroundColor: isDarkColor
                                    ? dynamicColor.withOpacity(0.5)
                                    : dynamicColor.withAlpha(100),
                              ),
                            ),
                          )
                        ],
                      ),
                    ),
                  ],
                )),
          );
        },
      ),
    );
  }
}


example/swipper_example/linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "swipper_example");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "swipper_example");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


example/swipper_example/windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.Create(L"swipper_example", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


example/swipper_example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:swipperview/swipperview.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: false,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<Map<String, dynamic>> productList = [
    {
      'title': '#1 Product of the Day',
      'description':
          '#1 This project is a starting point for a Dart package, a library module containing code that can be shared easily across multiple Flutter or Dart projects.',
      'url':
          'https://images.unsplash.com/photo-1611930022073-b7a4ba5fcccd?ixid=MnwxMjA3fDB8MHxzZWFyY2h8M3x8cHJvZHVjdHN8ZW58MHwxfDB8fA%3D%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=800&q=60',
    },
    {
      'title': '#2 Product of the Day',
      'description':
          '#2 This project is a starting point for a Dart package, a library module containing code that can be shared easily across multiple Flutter or Dart projects.',
      'url':
          'https://images.unsplash.com/photo-1610824352934-c10d87b700cc?ixid=MnwxMjA3fDB8MHxzZWFyY2h8MjV8fHByb2R1Y3RzfGVufDB8MXwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=800&q=60',
    },
    {
      'title': '#3 Product of the Day',
      'description':
          '#3 This project is a starting point for a Dart package, a library module containing code that can be shared easily across multiple Flutter or Dart projects.',
      'url':
          'https://images.unsplash.com/photo-1525966222134-fcfa99b8ae77?ixid=MnwxMjA3fDB8MHxzZWFyY2h8MzJ8fHByb2R1Y3RzfGVufDB8MXwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=800&q=60',
    },
    {
      'title': '#4 Product of the Day',
      'description':
          '#4 This project is a starting point for a Dart package, a library module containing code that can be shared easily across multiple Flutter or Dart projects.',
      'url':
          'https://images.unsplash.com/photo-1556228720-195a672e8a03?ixid=MnwxMjA3fDB8MHxzZWFyY2h8NTR8fHByb2R1Y3RzfGVufDB8MXwwfHw%3D&ixlib=rb-1.2.1&auto=format&fit=crop&w=800&q=60',
    },
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      extendBodyBehindAppBar: true,
      appBar: AppBar(
        elevation: 0,
        backgroundColor: Colors.transparent,
      ),
      body: Swipper(
        data: productList,
        icon: Icons.headphones,
        onChange: (page) {
          print('swipped $page');
        },
        actionName: 'Buy Now',
        actionHandler: (index) {
          print('action Click $index');
        },
      ),
    );
  }
}
